File: ftw_backend/src/models/category.py
from pydantic import BaseModel
from .counterpart import CounterpartView, CounterpartEdit, CounterpartFromId
from .category_type import CategoryType

# ======================================================================================================== #
#                                       BASE CLASSES
# ======================================================================================================== #

class Category(BaseModel):
    id: int = None
    owner_id: int = None

    name: str = ""
    description: str = ""
    category_type: CategoryType
    counterparts: list[CounterpartView] = []

    # Ensures we can easily convert from schema to model
    model_config = {'from_attributes': True}

class CategoryCreate(BaseModel):
    owner_id: int = None

    name: str
    description: str = ""
    category_type: str | CategoryType

    counterparts: list[CounterpartEdit] = []

    # Ensures we can easily convert from schema to model
    model_config = {'from_attributes': True}

class CategoryView(BaseModel):
    id: int
    name: str
    description: str = ""
    category_type: CategoryType
    counterparts: list[CounterpartView] = []

    # Ensures we can easily convert from schema to model
    model_config = {'from_attributes': True}

class CategoryEdit(BaseModel):
    id: int

    name: str
    description: str = None
    category_type: CategoryType
    
    counterparts: list[CounterpartFromId] = []

    # Ensures we can easily convert from schema to model
    model_config = {'from_attributes': True}

File: ftw_frontend/src/assets/types/Category.ts
import type { CategoryType, CategoryTypeEdit } from "./CategoryType";
import type { Counterpart, CounterpartEdit } from "./Counterpart";

export class Category {
  id: number;

  name: string;
  description: string;
  category_type: CategoryType;

  counterparts: Counterpart[];

  constructor(
    id: number,
    name: string,
    description: string,
    category_type: CategoryType,
    counterparts: Counterpart[],
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.category_type = category_type;
    this.counterparts = counterparts;
  }
}

export class CategoryEdit {
  id: number;
  name: string;
  description: string;
  category_type: CategoryTypeEdit;
  counterparts: CounterpartEdit[];

  constructor(
    id: number,
    name: string,
    description: string,
    category_type: CategoryTypeEdit,
    counterparts: CounterpartEdit[],
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.category_type = category_type;
    this.counterparts = counterparts;
  }
}
